src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs

```diff
@@ -164,24 +162,57 @@ public static DiagnosticAnalysisResult CreateFromBuilder(DiagnosticAnalysisResul
         // make sure we don't return null
         public ImmutableHashSet<DocumentId> DocumentIdsOrEmpty => DocumentIds ?? ImmutableHashSet<DocumentId>.Empty;
 
-        // this shouldn't be called for aggregated form.
-        public ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>> SyntaxLocals => ReturnIfNotDefault(_syntaxLocals);
-        public ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>> SemanticLocals => ReturnIfNotDefault(_semanticLocals);
-        public ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>> NonLocals => ReturnIfNotDefault(_nonLocals);
-        public ImmutableArray<DiagnosticData> Others => ReturnIfNotDefault(_others);
+        private ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>? GetMap(AnalysisKind kind)
+            => kind switch
+            {
+                AnalysisKind.Syntax => _syntaxLocals,
+                AnalysisKind.Semantic => _semanticLocals,
+                AnalysisKind.NonLocal => _nonLocals,
+                _ => throw ExceptionUtilities.UnexpectedValue(kind)
+            };
+
+        public ImmutableArray<DiagnosticData> GetAllDiagnostics()
+        {
+            // PERF: don't allocation anything if not needed
+            if (IsAggregatedForm || IsEmpty)
+            {
+                return ImmutableArray<DiagnosticData>.Empty;
+            }
+
+            Contract.ThrowIfNull(_syntaxLocals);
+            Contract.ThrowIfNull(_semanticLocals);
+            Contract.ThrowIfNull(_nonLocals);
+            Contract.ThrowIfTrue(_others.IsDefault);
+
+            return _syntaxLocals.Values.SelectMany(v => v).Concat(
```

Comment:
Nit: Might be better to avoid all the linq here and just use iterate and add to an array builder.

Reply:
Yeah, we could do that. I haven't changed it.